cpp/modmesh/serialization/SerializableItem.hpp

```diff
@@ -55,65 +56,265 @@ namespace detail
 /// Escape special characters in a string.
 std::string escape_string(std::string_view str_view);
 
-template <typename T>
-std::string to_json_string(const T & value)
+/// Trim leading and trailing whitespaces and control characters.
+std::string trim_string(const std::string & str);
+
+/// State of the JSON parser.
+enum class JsonState
+{
+    Start,
+    ObjectKey,
+    Column,
+    Comma,
+    ObjectValue,
+    End,
+}; /* end enum class JsonState */
+
+/// Type of JSON token.
+enum class JsonType
+{
+    Object,
+    Array,
+    String,
+    Number,
+    Boolean,
+    Null,
+    Unknown,
+}; /* end enum class JsonType */
+
+struct JsonNode; // Forward declaration
+typedef std::unique_ptr<JsonNode> JsonNodePtr;
+typedef std::unordered_map<std::string, JsonNodePtr> JsonMap;
+typedef std::vector<JsonNodePtr> JsonArray;
+
+struct JsonNode
 {
-    if constexpr (std::is_base_of_v<SerializableItem, T>)
+    using JsonValue = std::variant<JsonMap, JsonArray, std::string>;
+
+    JsonType type;
+    JsonValue value;
+
+    JsonNode(JsonType type, const std::string & expression)
+        : type(type)
     {
-        return value.to_json(); /* recursive here */
+        parse(expression);
     }
-    else if constexpr (std::is_convertible_v<T, std::string>)
+
+private:
+
+    void parse(const std::string & expression)
     {
-        return "\"" + escape_string(value) + "\"";
+        if (type == JsonType::Object)
```

Comment:
The function body of `parse` is a little bit too long for declaration, but it's fine for now.